### PR TITLE
torch.utils.checkpoint.checkpoint + torch.cuda.amp 

### PR DESCRIPTION
**Pull request by tano297**
Simple 2 line workaround to allow gradient checkpointing to work with amp autocast.
In the same way pytorch stores the "has_cuda" state in the context, we store a "has_autocast" during the first forward pass, so that we can re-enable it when the forward pass runs for the second time during the backward pass.

For anybody having this problem, the simple solution to this problem before this is merged can be found either here: #37730, or by simply copying this version of the file into your own codebase with the added 2 lines.

### CODE REVIEW

**Comment by zou3519** [comment]
@mcarilli I requested review from you because of the mention of amp, please let me know if that's not right

**Comment by tano297** [comment]
Any update on this? Let me know how I can help

**Comment by ZhichengHuang** [comment]
Is this bug solved? I have met the same issue.

**Comment by mcarilli** [comment]
Looks good to me, thanks!

I think the usual [custom autograd function decorators](https://pytorch.org/docs/stable/notes/amp_examples.html#functions-with-multiple-inputs-or-autocastable-ops) aren't preferable here, because `CheckpointFunction.backward` runs a nested forward and backward.  The autocast API recommends running only forward under autocast, but globally enabling autocast for all of `CheckpointFunction.backward` (as `@custom_bwd` might do) would include the nested backward as well.

Definitely needs a test though.

**Review by mcarilli — CHANGES_REQUESTED** [review]
Looks good to me, thanks!

I think the usual [custom autograd function decorators](https://pytorch.org/docs/stable/notes/amp_examples.html#functions-with-multiple-inputs-or-autocastable-ops) aren't preferable here, because `CheckpointFunction.backward` runs a nested forward and backward.  The autocast API recommends running only forward under autocast, but globally enabling autocast for all of `CheckpointFunction.backward` (as `@custom_bwd` might do) would include the nested backward as well.

Definitely needs a test though.

**Comment by mcarilli** [comment]
PR appears orphaned, moving to https://github.com/pytorch/pytorch/pull/49757.
